discretize/base/base_mesh.py

```diff
@@ -33,103 +31,164 @@ class BaseMesh(properties.HasProperties, InterfaceMixins):
+            R = np.atleast_2d(np.asarray(value, dtype=np.float64))
+            dim = self.dim
+            if R.shape != (dim, dim):
+                raise ValueError(
+                    f"Orientation matrix must be square and of shape {(dim, dim)}, got {R.shape}"
+                )
+            # Ensure each row is unitary
+            R = R / np.linalg.norm(R, axis=1)[:, None]
+            # Check if matrix is orthogonal
+            if not np.allclose(R @ R.T, np.identity(self.dim), rtol=1.e-5, atol=1E-6):
+                raise ValueError("Orientation matrix is not orthogonal")
+            self._orientation = R
 
-    @properties.validator("origin")
-    def _check_origin(self, change):
-        if not (
-            not isinstance(change["value"], properties.utils.Sentinel)
-            and change["value"] is not None
-        ):
-            raise Exception("n must be set prior to setting origin")
+    @property
+    def reference_system(self):
+        "The type of coordinate reference frame. Can take on the values "
+        return self._reference_system
 
-        if len(self._n) != len(change["value"]):
-            raise Exception(
-                "Dimension mismatch. origin has length {} != len(n) which is "
-                "{}".format(len(self.origin), len(self._n))
+    @reference_system.setter
+    def reference_system(self, value):
+        """Check if the reference system is of a known type."""
+        choices = ["cartesian", "cylindrical", "spherical"]
+        # Here are a few abbreviations that users can harnes
+        abrevs = {
+            "car": choices[0],
+            "cart": choices[0],
+            "cy": choices[1],
+            "cyl": choices[1],
+            "sph": choices[2],
+        }
+        # Get the name and fix it if it is abbreviated
+        value = value.lower()
+        value = abrevs.get(value, value)
+        if value not in choices:
+            raise ValueError(
+                "Coordinate system ({}) unknown.".format(self.reference_system)
             )
+        self._reference_system = value
+
+    def to_dict(self):
+        cls = type(self)
+        out = {
+            '__module__': cls.__module__,
+            '__class__': cls.__name__,
+        }
+        for item in self._items:
+            attr = getattr(self, item, None)
+            if attr is not None:
+                if isinstance(attr, np.ndarray):
+                    attr = attr.tolist()
+                elif isinstance(attr, tuple):
+                    # change to a list and make sure inner items are not numpy arrays
+                    attr = list(attr)
+                    for i, thing in enumerate(attr):
+                        if isinstance(thing, np.ndarray):
+                            attr[i] = thing.tolist()
+                out[item] = attr
+        return out
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+        return self.to_dict() == other.to_dict()
```

Comment:
I think it is great to implement an `__eq__`, but the dict-comparison is flaky with regards to, e.g., floating point issues.

I haven't tested it, but I suspect the following will return `False` instead of `True`:
```
d1 = discretize.TensorMesh([[0.3,]], (0,))
d2 = discretize.TensorMesh([[0.2+0.1,]], (0,))
d1 == d2
```